test/subscription_map.h

```diff
@@ -0,0 +1,265 @@
+//
+// Created by wkl04 on 10-12-2019.
+//
+
+#ifndef MQTTSUBSCRIPTION_SUBSCRIPTION_MAP_H
+#define MQTTSUBSCRIPTION_SUBSCRIPTION_MAP_H
+
+#include <mqtt/string_view.hpp>
+
+#include <unordered_map>
+#include <boost/functional/hash.hpp>
+
+#include "path_tokenizer.h"
+
+/**
+ *
+ * Non-normative comment
+        · $SYS/ has been widely adopted as a prefix to topics that contain Server-specific information or control APIs
+        · Applications cannot use a topic with a leading $ character for their own purposes
+
+    Non-normative comment
+        · A subscription to “#” will not receive any messages published to a topic beginning with a $
+        · A subscription to “+/monitor/Clients” will not receive any messages published to “$SYS/monitor/Clients”
+        · A subscription to “$SYS/#” will receive messages published to topics beginning with “$SYS/”
+        · A subscription to “$SYS/monitor/+” will receive messages published to “$SYS/monitor/Clients”
+        · For a Client to receive messages from topics that begin with $SYS/ and from topics that don’t begin with a $, it has to subscribe to both “#” and “$SYS/#”
+ */
+
+template<typename Value>
+class subscription_map_base
+{
+    using node_id = size_t;
+    using path_entry_key = std::pair< node_id, std::string>;
+
+    static constexpr node_id root_node_id = 0;
+
+    struct path_entry
+    {
+        node_id id;
+
+        bool has_hash_child : 1;
+        bool has_plus_child : 1;
+        uint32_t count : 30;
+
+        Value value;
+
+        path_entry(node_id _id)
+                : id(_id), has_hash_child(false), has_plus_child(false), count(1)
+        { }
+    };
+
+    using map_type = std::unordered_map< path_entry_key, path_entry, boost::hash< path_entry_key > >;
+    using map_type_iterator = typename map_type::iterator;
+    using map_type_const_iterator = typename map_type::const_iterator;
+
+    map_type map;
+    map_type_iterator root;
+    node_id next_node_id;
+
+protected:
+    map_type_iterator end() { return map.end(); }
+
+    std::vector< std::pair<map_type_iterator, map_type_iterator> > find_subscription(MQTT_NS::string_view const &topic)
```

Comment:
I'd rather see these functions take one of the following as the parameter:

1) Some kind of wrapper object that automatically parses the underlying string data as needed.
2) Some object that represents a pre-tokenized form of the topic.

From the brokers perspective, it only wants to be tokenizing things once. So 2) Is useful for code patterns where the pre-tokenized form of the topic object is used in multiple different places.

But I think it's important that we try to aim for a situation where the broker can, when it needs to, immediately grab a handle to the actual subscription object one time, and then refer to the subscription object that it has a handle for. That way tokenization should only ever happen one time per request from a client.

Reply:
The tokenization already happens once. Once a client subscribes to a topic, the subscription is tokenized and stored in the subscription map. This is stored until the client disconnects or unsubscribes. 

Once a topic is published, it needs to be tokenized and checked which subscriptions match. This has to be redone for every publish message, unless it is some stored topic/topic alias from mqttv5, there you can save some tokenization. At the cost of keeping the tokens in memory. The tokenization is very cheap, I think you would rather store a single string, then many fragmented strings on the heap.

Reply:
The way you've designed this data structure, the tokenization happens any time someone calls 
 any of these functions.

My concern was that as a consumer of the API of this data structure, i may need to call several API functions from your datastructure within the same function call. The way you've set your API up, if I did that, I pay the cost of re-tokenizing every time.

Reply:
Yes, but the tokenizing is very efficient now. No memory allocation is done. If you would store the intermediate parsed tokens, this would require a memory allocation. Which is probably worse in performance.

Reply:
We can satisfy my concern if there's is some way for me to acquire a handle, e.g. an iterator, to the item inside the data structure that's associated with the post-tokenized topic.

is this possible?

Reply:
It is possible, but why would you want to do that ? I timed the two approaches, see the code here:

https://pastebin.com/yHaAwqKi

The results are as follows:

Sum: 1200000000
Stored milliseconds: 21651
Sum: 1200000000
Direct milliseconds: 2676

Using a pretokenized approach is about 8x slower than tokenizing and using the substrings immediatly. This is considering I tokenize and use the tokenized results 10.000x times. I could not find a scenario in which pre-tokenizing would be faster. So why do you want to do that ?

Reply:
I think maybe you misunderstood my question.

The question was to be able to do this:

```
auto handle = map.get_handle("asdfasdf"); <- Tokenization happens
map.use_handle(handle); <- Tokenization doesn't happen
```

Reply:
You mean you request a handle to a certain subscription, and then you use the handle to update the entry or delete the entry ?

Reply:
Without paying the overhead or having to re-search the entry in the map

Reply:
I think I understand what you mean, and yes, this can be benefit in performance, to avoid having to do the tree lookup.

Reply:
Right, that's what I meant.

I just wanted to be able to get a handle to the entry in the map, and then use that handle to access data from it / update it.

I don't know that the code you've written to integrate with the test_broker code would use that right now, but having that flexibility would be useful.

Reply:
I looked at this, and it is possible, but it only makes sense for an update and multiple subscription erase. You can store the key of the leaf entry, and it allows you to update the value, without having to  lookup the complete path of the subscription. For an erase in the multiple subscription map it has some benefit, if multiple values are subscribed to the same path, you can use the leaf entry to remove it from the set.

Reply:
I added the possibility to use a handle to update / remove an entry in the subscription map. At the moment the test broker only uses it to remove the entry from the subscription map. It could use it to update the qos when the entry already exists.